clang/lib/3C/AVarBoundsInfo.cpp

```diff
@@ -280,28 +301,21 @@ bool AvarBoundsInference::getReachableBoundKeys(const ProgramVarScope *DstScope,
     PotK.insert(FromVarK);
   }
 
-  // Get all bounds key that are equivalent to FromVarK
-  std::set<BoundsKey> AllFKeys;
-  AllFKeys.clear();
-  AllFKeys.insert(FromVarK);
-
-  for (auto CurrVarK : AllFKeys) {
-    // Find all the in scope variables reachable from the CurrVarK
-    // bounds variable.
-    std::set<BoundsKey> InScopeKeys;
-    std::set<BoundsKey> VisibleKeys;
-    if (DstScope->isInInnerScope(*BI->getProgramVar(CurrVarK)->getScope()))
-      VisibleKeys.insert(CurrVarK);
-    ScopeVisitor TV(DstScope, InScopeKeys, VisibleKeys, BI->PVarInfo,
-                    BI->PointerBoundsKey);
-    BKGraph.visitBreadthFirst(CurrVarK,
-                              [&TV](BoundsKey BK) { TV.visitBoundsKey(BK); });
-    // Prioritize in scope keys.
-    if (!InScopeKeys.empty()) {
-      PotK.insert(InScopeKeys.begin(), InScopeKeys.end());
-    } else {
-      PotK.insert(VisibleKeys.begin(), VisibleKeys.end());
-    }
+  // Find all in scope variables reachable from the FromVarK bounds variable.
+  ScopeVisitor TV(DstScope, BI->PVarInfo, BI->PointerBoundsKey);
+  BKGraph.visitBreadthFirst(FromVarK,
+                            [&TV](BoundsKey BK) { TV.visitBoundsKey(BK); });
+  // Prioritize in scope keys.
+  if (!TV.getInScopeKeys().empty()) {
+    PotK.insert(TV.getInScopeKeys().begin(), TV.getInScopeKeys().end());
+  } else {
+    PotK.insert(TV.getVisibleKeys().begin(), TV.getVisibleKeys().end());
+
+    // This condition seems to be necessary for array bounds using global
+    // variables, but it's not clear why exactly it's required, and why it's
+    // only added to VisibleKeys and never InScopeKeys.
+    if (DstScope->isInInnerScope(*BI->getProgramVar(FromVarK)->getScope()))
+      PotK.insert(FromVarK);
```

Comment:
I do not think this is needed. Given that we do this check in `ScopeVisitor`. Can you remove this and run tests to confirm?

Reply:
I looked into this more, and it seems that global variables aren't ever present in the `BKGraph` parameter, so they're never visited by the breadth first search. This makes sure they can still be considers as potential bounds keys.